tests/lib/snaps/test-snapd-eds/calendar.c

```diff
@@ -0,0 +1,201 @@
+    struct open_data data = {
+        .main_loop = main_loop,
+        .source_id = source_id,
+        .error = error,
+        .calendar = &calendar,
+        .should_quit = FALSE,
+    };
+    guint source_added_id = g_signal_connect(registry, "source-added",
+                                             G_CALLBACK(source_added), &data);
+
+    // Create a new local calendar with the desired source ID
+    scratch = e_source_new_with_uid(source_id, NULL, error);
+    if (!scratch)
+        goto end;
+    e_source_set_display_name(scratch, source_id);
+    ESourceBackend *backend = e_source_get_extension(
+        scratch, E_SOURCE_EXTENSION_CALENDAR);
+    e_source_backend_set_backend_name(backend, "local");
+
+    // Try to commit the new source to the registry, which will fail
+    // if it already exists
+    if (!e_source_registry_commit_source_sync(
+            registry, scratch, NULL, &commit_error)) {
+        if (g_error_matches(commit_error, G_IO_ERROR, G_IO_ERROR_EXISTS)) {
+            g_autoptr(ESource) source = e_source_registry_ref_source(
+                registry, source_id);
+            source_added(registry, source, &data);
+        } else {
+            g_propagate_error(error, commit_error);
+            commit_error = NULL;
+            goto end;
+        }
+    }
+
+    // If we don't have the calendar at this point, wait on the
+    // source-added signal for it to be created.  Set a timer so we
+    // don't wait forever.
+    if (!calendar) {
+        guint timeout_id = g_timeout_add_seconds(
+            20, source_added_timeout, &data);
+        data.should_quit = TRUE;
+        g_main_loop_run (main_loop);
+        g_source_remove(timeout_id);
+    }
+
+end:
+    if (source_added_id) {
+        g_signal_handler_disconnect(registry, source_added_id);
+    }
+    return g_steal_pointer(&calendar);
+}
+
+static gboolean
+load_event_from_stdin(ECalClient *calendar, GError **error)
+{
+    g_autoptr(GString) ics_data = g_string_new(NULL);
+    char buffer[4092];
+    size_t n_read = fread(buffer, 1, sizeof(buffer), stdin);
+    while (n_read > 0) {
+        g_string_append_len(ics_data, buffer, n_read);
+        n_read = fread(buffer, 1, sizeof(buffer), stdin);
+    }
+
+    g_autoptr(ECalComponent) component = e_cal_component_new_from_string(ics_data->str);
+    if (!component) {
+        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+                            "could not parse iCalendar data");
+        return FALSE;
+    }
+
+    return e_cal_client_create_object_sync(
+        calendar, e_cal_component_get_icalcomponent(component), NULL,
+        NULL, error);
+}
+
+static gboolean
+list_events(ECalClient *calendar, GError **error)
+{
+    GSList *results = NULL;
+    if (!e_cal_client_get_object_list_as_comps_sync(calendar, "#t", &results,
+                                                    NULL, error)) {
+        return FALSE;
+    }
+
+    const GSList *l;
```

Comment:
'l' as a variable name is a bit hard to read. I suggest 'lst' or some other character. Not a blocker.

Reply:
I'm not sure that would really improve readability much: this is pretty much the equivalent of using `i` as a variable when iterating over an array: it's use is localised over two lines, and the more important variables have more meaningful names (the list head, and the data member of the linked list).